### PR TITLE
add protocol 300 support

### PR DESCRIPTION
**Pull request by jmue**
This pull request adds protocol 300 support to your library. I've tested my changes successfully with a VScotHO1 heater. But as I am unfamiliar with rust and its specific patterns I ask you to review this piece of code carefully.

### CODE REVIEW

File: src/protocol/kw2.rs [hunk 1]
```diff
@@ -16,9 +16,6 @@ impl Kw2 {
     let start = Instant::now();
 
     loop {
-      o.write_all(&[0x04])?;
-      o.flush()?;
```
Comment (reitermarkus) [inline src/protocol/kw2.rs:20]: Are you sure this isn't needed? As far as I can remember, it *does* work without this, but the response time is lower if you send `0x04` before waiting for `0x05`.

File: src/protocol/p300.rs [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+    if [RESPONSE, WRITEDATA] != write_response[0..2] {
+      return Err(io::Error::new(io::ErrorKind::InvalidData, "invalid write data response"));
+    }
+    if *addr != write_response[2..4] {
+      return Err(io::Error::new(io::ErrorKind::InvalidData, "wrong address"));
+    }
+    if value.len() as u8 != write_response[5] {
+      return Err(io::Error::new(
+        io::ErrorKind::InvalidData,
+        "could not write data",
```
Comment (reitermarkus) [inline src/protocol/p300.rs]: Make this a single line.

File: src/protocol/p300.rs [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+
+    Self::write_telegram(o, &write_request)?;
+
+    let write_response = Self::read_telegram(o)?;
+
+    let expected_len: usize = 5;
+    if expected_len != write_response.len() {
+      return Err(io::Error::new(
+        io::ErrorKind::InvalidData,
+        "unexpected response length",
```
Comment (reitermarkus) [inline src/protocol/p300.rs]: Make this a single line.

File: src/protocol/p300.rs [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+      return Err(io::Error::new(io::ErrorKind::InvalidData, "invalid read data response"));
+    }
+    if *addr != read_response[2..4] {
+      return Err(io::Error::new(io::ErrorKind::InvalidData, "wrong address"));
+    }
+    if buf.len() as u8 != read_response[4] {
+      return Err(io::Error::new(io::ErrorKind::InvalidData, "wrong data length"));
+    }
+
+	buf.clone_from_slice(&read_response[5..(5 + buf.len())]);
```
Comment (reitermarkus) [inline src/protocol/p300.rs]: Fix the indentation here.

File: src/protocol/p300.rs [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+
+    Self::write_telegram(o, &read_request)?;
+
+    let read_response = Self::read_telegram(o)?;
+
+    let expected_len: usize = 5 + buf.len();
+    if expected_len != read_response.len() {
+      return Err(io::Error::new(
+        io::ErrorKind::InvalidData,
+        "unexpected response length",
```
Comment (reitermarkus) [inline src/protocol/p300.rs]: Make this a single line.

File: src/protocol/p300.rs [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+
+      let stop = Instant::now();
+      if (stop - start) > Optolink::TIMEOUT {
+        break;
+      }
+    }
+
+    Err(io::Error::new(
+      io::ErrorKind::TimedOut,
+      "send telegram timed out",
```
Comment (reitermarkus) [inline src/protocol/p300.rs]: Make this a single line.

File: src/protocol/p300.rs [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+
+      o.read_exact(&mut buf)?;
+      let data_length: usize = buf[0] as usize;
+
+      let mut data = vec![0; data_length];
+      o.read_exact(&mut data)?;
+      let data = data;
+
+      let checksum: u8 = data.iter().sum();
+	  let checksum = checksum + data_length as u8;
```
Comment (reitermarkus) [inline src/protocol/p300.rs]: Fix the indentation here.

File: src/protocol/p300.rs [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+
+      let mut status = [0xff];
+      o.read_exact(&mut status)?;
+      match status {
+        [ACK] => return Ok(()),
+        [NACK] => {}
+        [_] => {
+          return Err(io::Error::new(
+            io::ErrorKind::InvalidData,
+            "send telegram failed",
```
Comment (reitermarkus) [inline src/protocol/p300.rs]: Make this a single line.

File: src/protocol/p300.rs [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+const WRITEDATA: u8 = 0x02;
+//const FUNCTIONCALL: u8 = 0x07;
+
+#[derive(Debug)]
+pub struct P300;
+
+impl P300 {
+  #[inline]
+  fn write_telegram(o: &mut Optolink, data: &[u8]) -> Result<(), std::io::Error> {
+	let data_length = data.len() as u8;
```
Comment (reitermarkus) [inline src/protocol/p300.rs]: Fix the indentation here.

File: src/protocol/p300.rs [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+    let checksum: u8 = data.iter().sum();
+	let checksum = checksum + data_length;
```
Comment (reitermarkus) [inline src/protocol/p300.rs]: Fix the indentation here.

File: src/protocol/p300.rs [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+    let mut buf = [0xff];
+
+    let start = Instant::now();
+
+    loop {
+      o.read_exact(&mut buf)?;
+      if buf != [LEADIN] {
+        return Err(io::Error::new(
+          io::ErrorKind::InvalidData,
+          "telegram leadin expected",
```
Comment (reitermarkus) [inline src/protocol/p300.rs]: Make this a single line.

File: src/protocol/p300.rs [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+
+      let stop = Instant::now();
+      if (stop - start) > Optolink::TIMEOUT {
+        break;
+      }
+    }
+
+    Err(io::Error::new(
+      io::ErrorKind::TimedOut,
+      "send telegram timed out",
```
Comment (reitermarkus) [inline src/protocol/p300.rs]: Make this a single line.

File: src/protocol/p300.rs [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+use std::time::Instant;
+
+use crate::Optolink;
+
+use super::Protocol;
+
+const LEADIN: u8    = 0x41;
+
+const RESET: u8     = 0x04;
+const SYNC: u8      = 0x05;
```
Comment (reitermarkus) [inline src/protocol/p300.rs:11]: Could you also refactor `Kw2` to use a `SYNC` constant?

**Review by reitermarkus — CHANGES_REQUESTED** [review]
Thanks for the PR, great work. I'll take another look once the style issues are fixed.

File: src/protocol/kw2.rs (hunk 1 above)
Comment (jmue) [reply src/protocol/kw2.rs:20]: From my understandings of the [documentaion](https://github.com/openv/openv/wiki/Protokoll-300)  the '0x04' resets the Optolink connection. The trick is to the reset the Optolink before every command to get a faster '0x05' (SYNC), right? Ok, that makes sense. I will undo it. But should then the 'negotiate' function send another '0x04', or not?

Comment (reitermarkus) [reply src/protocol/kw2.rs:20]: > The trick is to the reset the Optolink before every command to get a faster '0x05' (SYNC), right?

Yes, exactly.

> But should then the 'negotiate' function send another '0x04', or not?

I guess it doesn't hurt to send `0x04` when initializing, so I would suggest keeping the `0x04` in `negotiate` and then re-using `negotiate` inside `sync`.

File: src/protocol/p300.rs [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+      if checksum == buf[0] {
+        o.write_all(&[ACK])?;
+        o.flush()?;
+        return Ok(message)
+      }
+
+      o.write_all(&[NACK])?;
+      o.flush()?;
+
+      let stop = Instant::now();
```
Comment (reitermarkus) [inline src/protocol/p300.rs:90]: Add a newline after this.

File: src/protocol/p300.rs [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+const LEADIN: u8    = 0x41;
+
+const RESET: u8     = 0x04;
+const SYNC: u8      = 0x05;
+const ACK: u8       = 0x06;
+const NACK: u8      = 0x15;
+
+const REQUEST: u8   = 0x00;
+const RESPONSE: u8  = 0x01;
+//const ERROR: u8    = 0x03;
```
Comment (reitermarkus) [inline src/protocol/p300.rs]: When would this occur?

File: src/protocol/p300.rs [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+
+const READDATA: u8  = 0x01;
+const WRITEDATA: u8 = 0x02;
+//const FUNCTIONCALL: u8 = 0x07;
```
Comment (reitermarkus) [inline src/protocol/p300.rs]: When would this be used?

File: src/protocol/p300.rs [hunk 17]
```diff
@@ -0,0 +1,188 @@
...
+
+      let mut status = [0xff];
+      o.read_exact(&mut status)?;
+      match status {
+        [ACK] => return Ok(()),
+        [NACK] => {}
+        [_] => return Err(io::Error::new(io::ErrorKind::InvalidData, "send telegram failed")),
+      }
+
+      let stop = Instant::now();
```
Comment (reitermarkus) [inline src/protocol/p300.rs:48]: Add a newline after this.

File: src/protocol/p300.rs [hunk 18]
```diff
@@ -0,0 +1,188 @@
...
+    loop {
+      o.read_exact(&mut status)?;
+      match status {
+        [ACK] => return Ok(()),
+        [NACK] => {}
+        [SYNC] => {
+          o.write_all(&[0x16, 0x00, 0x00])?;
+          o.flush()?;
+        }
+        _ => return Err(io::Error::new(io::ErrorKind::InvalidData, "UNKNOWN")),
```
Comment (reitermarkus) [inline src/protocol/p300.rs]: Maybe a better error message here?

File: src/protocol/kw2.rs (hunk 1 above)
Comment (reitermarkus) [reply src/protocol/kw2.rs:20]: And maybe add a comment explaining that this is in order to get a faster `SYNC` response.

**Comment by jmue** [comment]
I modified the 'negotiate' function for p300 slightly as the optolink reset request [is | is not] acknowledged depending on the previous state.

File: src/protocol/p300.rs (hunk 16 above)
Comment (jmue) [reply src/protocol/p300.rs]: P300 specifies remote function calls for example reset_control_panel_heater_settings. I removed this untils this comes into place.

File: src/protocol/p300.rs (hunk 15 above)
Comment (jmue) [reply src/protocol/p300.rs]: I dont know really. I think of a [ERROR, READ] or [ERROR, WRITE] response. But I have not seen this. Dead code is bad code, so I removed this.

File: src/vcontrol.rs [hunk 19]
```diff
@@ -7,7 +7,8 @@ pub struct VControl<D: Device> {
 }
 
 impl<D: Device> VControl<D> {
-  pub fn new(device: Optolink) -> Self {
+  pub fn new(mut device: Optolink) -> Self {
+    D::Protocol::negotiate(&mut device).expect("protocol negotiation failed");
```
Comment (reitermarkus) [inline src/vcontrol.rs]: Maybe we should rename this function to `connect` and return a `Result` instead?

Comment (jmue) [reply src/vcontrol.rs]: Yes, I will have a look on that.

Comment (jmue) [reply src/vcontrol.rs]: Done.

**Comment by reitermarkus** [comment]
Thanks, @jmue, nice work!

**Pull request merged**